lib/fluent/plugin/out_forward/ack_handler.rb

```diff
@@ -0,0 +1,161 @@
+#
+# Fluentd
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'fluent/plugin/output'
+require 'fluent/plugin_helper/socket'
+require 'fluent/engine'
+require 'fluent/clock'
+
+module Fluent::Plugin
+  class ForwardOutput < Output
+    class AckHandler
+      module Result
+        SUCCESS = :success
+        FAILED = :failed
+        CHUNKID_UNMATCHED = :chunkid_unmatched
+      end
+
+      def initialize(timeout:, log:, read_length:)
+        @mutex = Mutex.new
+        @ack_waitings = []
+        @timeout = timeout
+        @log = log
+        @read_length = read_length
+        @unpacker = Fluent::Engine.msgpack_unpacker
+      end
+
+      def collect_response(select_interval)
+        now = Fluent::Clock.now
+        sockets = []
+        results = []
+        begin
+          new_list = []
+          @mutex.synchronize do
+            @ack_waitings.each do |info|
+              if info.expired?(now)
+                # There are 2 types of cases when no response has been received from socket:
+                # (1) the node does not support sending responses
+                # (2) the node does support sending response but responses have not arrived for some reasons.
+                @log.warn 'no response from node. regard it as unavailable.', host: info.node.host, port: info.node.port
+                results << [info, Result::FAILED]
+              else
+                sockets << info.sock
+                new_list << info
+              end
+            end
+            @ack_waitings = new_list
+          end
+
+          readable_sockets, _, _ = IO.select(sockets, nil, nil, select_interval)
+          if readable_sockets
+            readable_sockets.each do |sock|
+              results << read_ack_from_sock(sock)
+            end
+          end
+
+          results.each do |info, ret|
+            if info.nil?
+              yield nil, nil, nil, ret
+            else
+              yield info.chunk_id, info.node, info.sock, ret
+            end
+          end
+        rescue => e
+          @log.error 'unexpected error while receiving ack', error: e
+          @log.error_backtrace
+        end
+      end
+
+      ACKWaitingSockInfo = Struct.new(:sock, :chunk_id, :chunk_id_base64, :node, :time, :timeout) do
```

Comment:
I don't remember why this struct has `time` and `timeout` fields and calculate expired time in `expired?` call.
Could we change this struct like below?

```
ACKWaitingSockInfo = Struct.new(:sock, :chunk_id, :chunk_id_base64, :node, :expired_time) do
  def expired?(now)
    expired_time < now
  end
end
```